### ISSUE TITLE
"Test timed out" on benchmark_dataflow_lstm

### ISSUE THREAD

**Issue by gchaperon** [ChrisCummins/ProGraML#104]
Hi! I was trying to run the tests following the intruction in the README but  i'm having a timeout error.

I'm doing the following steps in a aws ec2 instace with t4 gpu, ubuntu 18.04 and a clean python3.6 venv, using the stable branch:
1. install bazel from apt repo, following the instructions from [here](https://docs.bazel.build/versions/3.4.0/install-ubuntu.html)
2. step 2 from `DEPS.md` file
3. upgrade pip with `pip install --upgrade pip`, which bumps it to 20.2.2
4. step 3 from `DEPS.md` file
5. `pip install tensorflow-gpu==1.14.0` (before this the gpu wasn't recognized)
6. run `tf.test.is_gpu_available()` in interpreter, returns `True`
7. run `bazel test //programl/...`

After a couple of minutes I get
```INFO: Elapsed time: 1590.389s, Critical Path: 972.54s
INFO: 1171 processes: 1171 linux-sandbox.
INFO: Build completed, 1 test FAILED, 1602 total actions
//programl/graph/analysis:datadep_test                                   PASSED in 21.4s
//programl/graph/analysis:dominance_test                                 PASSED in 51.0s
//programl/graph/analysis:liveness_test                                  PASSED in 0.1s
//programl/graph/analysis:reachability_test                              PASSED in 20.5s
//programl/graph/analysis:subexpressions_test                            PASSED in 12.9s
//programl/graph/format:cdfg_test                                        PASSED in 0.1s
//programl/graph/format:graph_serializer_test                            PASSED in 0.1s
//programl/graph/format:graph_tuple_test                                 PASSED in 0.1s
//programl/graph/format:node_link_graph_test                             PASSED in 0.1s
//programl/graph/format/py:graph_serializer_test                         PASSED in 3.0s
//programl/graph/format/py:graph_tuple_builder_test                      PASSED in 3.5s
//programl/graph/format/py:nx_format_test                                PASSED in 4.3s
//programl/graph/py:program_graph_builder_test                           PASSED in 3.1s
//programl/ir/llvm:clang_test                                            PASSED in 0.4s
//programl/ir/llvm:inst2vec_encoder_test                                 PASSED in 35.2s
//programl/ir/llvm/py:llvm_test                                          PASSED in 3.1s
//programl/ir/xla:hlo_proto_reader_test                                  PASSED in 0.1s
//programl/ir/xla/py:xla_test                                            PASSED in 6.2s
//programl/models:rolling_results_test                                   PASSED in 5.9s
//programl/task/devmap/dataset:create_test                               PASSED in 91.8s
//programl/test/benchmarks:benchmark_dataflow_analyses                   PASSED in 122.2s
//programl/test/benchmarks:benchmark_dataflow_ggnn                       PASSED in 207.6s
//programl/test/benchmarks:benchmark_llvm2graph                          PASSED in 55.7s
//programl/test/benchmarks:benchmark_dataflow_lstm                      TIMEOUT in 900.3s
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/cebc236d29bba7692e479297d0d6d320/execroot/programl/bazel-out/k8-fastbuild/testlogs/programl/test/benchmarks/benchmark_dataflow_lstm/test.log

Executed 24 out of 24 tests: 23 tests pass and 1 fails locally.
INFO: Build completed, 1 test FAILED, 1602 total actions
```

The content from `.../benchmark_dataflow_lstm/test.log`

```exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //programl/test/benchmarks:benchmark_dataflow_lstm
-----------------------------------------------------------------------------

=== BENCHMARK 1: Loading graphs from filesystem ===
9620 graphs [00:29, 329.01 graphs/s]
I0826 07:14:40 prof.py:181] Benchmark graph loader in 29s 253ms

=== BENCHMARK 2: Batch construction ===
151 batches [01:54,  1.32 batches/s]
I0826 07:16:54 prof.py:181] Benchmark batch construction in 1m 54s 536ms

=== BENCHMARK 3: Model training ===
2020-08-26 07:16:54.433935: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2020-08-26 07:16:54.459252: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcuda.so.1
2020-08-26 07:16:55.261248: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.263185: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x2634adf0 executing computations on platform CUDA. Devices:
2020-08-26 07:16:55.263212: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): Tesla T4, Compute Capability 7.5
2020-08-26 07:16:55.285014: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2499995000 Hz
2020-08-26 07:16:55.285231: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x269272d0 executing computations on platform Host. Devices:
2020-08-26 07:16:55.285246: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
2020-08-26 07:16:55.285553: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.286528: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties: 
name: Tesla T4 major: 7 minor: 5 memoryClockRate(GHz): 1.59
pciBusID: 0000:00:1e.0
2020-08-26 07:16:55.289582: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.0
2020-08-26 07:16:55.348916: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0
2020-08-26 07:16:55.377464: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcufft.so.10.0
2020-08-26 07:16:55.386473: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcurand.so.10.0
2020-08-26 07:16:55.452863: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusolver.so.10.0
2020-08-26 07:16:55.495030: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusparse.so.10.0
2020-08-26 07:16:55.603611: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
2020-08-26 07:16:55.603800: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.604845: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.605713: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0
2020-08-26 07:16:55.606491: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.0
2020-08-26 07:16:55.608628: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-08-26 07:16:55.608645: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0 
2020-08-26 07:16:55.608651: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N 
2020-08-26 07:16:55.609510: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.610498: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.611392: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 14257 MB memory) -> physical GPU (device: 0, name: Tesla T4, pci bus id: 0000:00:1e.0, compute capability: 7.5)
2020-08-26 07:16:55.643593: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.644511: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties: 
name: Tesla T4 major: 7 minor: 5 memoryClockRate(GHz): 1.59
pciBusID: 0000:00:1e.0
2020-08-26 07:16:55.644555: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.0
2020-08-26 07:16:55.644568: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0
2020-08-26 07:16:55.644580: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcufft.so.10.0
2020-08-26 07:16:55.644591: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcurand.so.10.0
2020-08-26 07:16:55.644602: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusolver.so.10.0
2020-08-26 07:16:55.644614: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusparse.so.10.0
2020-08-26 07:16:55.644626: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
2020-08-26 07:16:55.644699: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.645613: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.646468: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0
2020-08-26 07:16:55.646487: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-08-26 07:16:55.646494: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0 
2020-08-26 07:16:55.646499: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N 
2020-08-26 07:16:55.646582: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.647484: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.648352: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/device:GPU:0 with 14257 MB memory) -> physical GPU (device: 0, name: Tesla T4, pci bus id: 0000:00:1e.0, compute capability: 7.5)
2020-08-26 07:16:55.790275: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.791246: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties: 
name: Tesla T4 major: 7 minor: 5 memoryClockRate(GHz): 1.59
pciBusID: 0000:00:1e.0
2020-08-26 07:16:55.791301: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.0
2020-08-26 07:16:55.791314: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0
2020-08-26 07:16:55.791325: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcufft.so.10.0
2020-08-26 07:16:55.791337: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcurand.so.10.0
2020-08-26 07:16:55.791349: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusolver.so.10.0
2020-08-26 07:16:55.791361: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusparse.so.10.0
2020-08-26 07:16:55.791373: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
2020-08-26 07:16:55.791450: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.792357: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.793240: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0
2020-08-26 07:16:55.793268: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-08-26 07:16:55.793275: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0 
2020-08-26 07:16:55.793281: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N 
2020-08-26 07:16:55.793387: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.794295: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-08-26 07:16:55.795170: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/device:GPU:0 with 14257 MB memory) -> physical GPU (device: 0, name: Tesla T4, pci bus id: 0000:00:1e.0, compute capability: 7.5)
Train:   0%|          | 0/192 [00:00<?, ? graphs/s, f1=?, loss=?, prec=?, rec=?]2020-08-26 07:17:02.668086: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0
2020-08-26 07:17:04.545515: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
Train: 100%|##########| 192/192 [00:14<00:00, 13.08 graphs/s, f1=0.377, loss=0.0244, prec=0.300, rec=0.529]
I0826 07:17:10 prof.py:181] Benchmark training (prebuilt batches) in 14s 683ms
Train: 192 graphs [00:19, 40.31 graphs/s, f1=0.336, loss=0.0234, prec=0.254, rec=0.500]-- Test timed out at 2020-08-26 07:28:51 UTC --
```

I see from [here](https://github.com/ChrisCummins/ProGraML/blob/stable/programl/test/benchmarks/benchmark_dataflow_lstm.py#L153) that there is still one benchmark left and I have no idea why it is getting stuck at that point.

Please feel free to ask for more logs, i'm kinda new to bazel and i don't know where else to look for.

Thanks!

PS: I didn' include the output from `bazel run //tools:whoami` because it seems it's not working in stable branch

**Comment by ChrisCummins** [comment]
Hi @gchaperon , thanks a lot for the super-detailed report! I will take a look at reproducing and posting a fix. I believe the issue is localized to the LSTM code. If you are not interested in running the LSTM baseline experiments, you can safely ignore this error.

Cheers,
Chris

**Comment by ChrisCummins** [comment]
The underlying issue here was #140 and a fix was merged in #142.

Cheers,
Chris
